### PR TITLE
feat: informative user-agent for otlp exporters

### LINKED ISSUE #2209: Collect OpenTelemetry collector telemetry data for service owners
**Is your feature request related to a problem? Please describe.**
As the OpenTelemetry collector maintainer and backend service owner, I want to have a way to collect the collector telemetry data so I can
* Get the adoption rate on each version of the released OpenTelemetry Collector that sends metrics/traces/logs data to its backend services
* Identify the running environment where OpenTelemetry collector is running. (Eg, for AWS customers, we want to tell where the collector is running. Eg, ECS, EKS, Lambda, etc.)
* Identify the customers who got impacted by the urgent issues(security, sever bugs, etc) that detected on a specific OpenTelemetry Collector version. So the service owner should be able to notify those customers. We can find the impacted customers by querying the customer IDs in service logs sent from OpenTelemetry Collector.

**Describe the solution you'd like**
When OpenTelemetry Collector sends traces/metrics/logs data to backend in the exporters, we can attach the telemetry data into User-Agent of the API requests in the exporters. 
For example,
**Set UserAgent to**: `OpenTelemetryCollector v0.1.0 go1.14 linux amd64 ECS`

**Implementation**,
Create a common module under `internal` folder  to construct the telemetry data content for all the exporters to use before sending the data to backend services.

### PR DESCRIPTION
**Pull request by vreynolds**
**Description:** This adds a default user-agent header to outgoing requests by the OTLP/gRPC and OTLP/HTTP exporters. The default user-agent contains build information about the collector.

Behavior details:
- Currently, it's possible to add a user-agent header to the HTTP exporter via the `headers` config. This change preserves this behavior, and only sets the collector user-agent, if one is not present in config headers.
- Currently, it's *not* possible to do the above with a gRPC exporter. This change does not affect that, and only updates the existing gRPC client user-agent.

**Link to tracking Issue:** Fixes #2209 

**Testing:** Updated/added to existing unit tests. Built a local collector and verified outgoing user-agent headers in both exporters.

**Documentation:** N/A?

### CODE REVIEW

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [inline config/confighttp/confighttp.go:66]: Changing the interface here doesn't feel right since the rest of the configuration happens by setting values on the `HTTPClientSettings` struct.  What about changing `addDefaultUserAgent()` to an exported method on that struct that can be invoked before `ToClient()`?

File: config/configgrpc/configgrpc.go
Comment (Aneurysm9) [inline config/configgrpc/configgrpc.go:183]: Similarly not sure about changing the interface here.  Can the `OTLPExporter` instead do the addition of the user agent option?

```go
dialOpts, err := config.GRPCClientSettings.ToDialOptions(ext)
dialOpts = append(dialOpts, grpc.WithUserAgent(exporterhelper.DefaultUserAgent(info)))
```

File: exporter/exporterhelper/common.go
Comment (Aneurysm9) [inline exporter/exporterhelper/common.go:224]: It might be nice to add `GOOS` and `GOARCH` here as well.

File: config/confighttp/confighttp.go
Comment (vreynolds) [reply config/confighttp/confighttp.go]: Updated in the new commit

File: config/configgrpc/configgrpc.go
Comment (vreynolds) [reply config/configgrpc/configgrpc.go]: Updated in the new commit

File: exporter/exporterhelper/common.go
Comment (vreynolds) [reply exporter/exporterhelper/common.go]: Updated in the new commit

**Review by Aneurysm9 — APPROVED** [review]

File: exporter/otlpexporter/otlp.go
Comment (codeboten) [inline exporter/otlpexporter/otlp.go:41]: I wonder if we should put `buildInfo` in the TelemetrySettings, any thoughts @bogdandrutu?

**Review by codeboten — COMMENTED** [review]
👍 this is a good feature to add. Just one question regarding the following in the description:

>    Currently, it's possible to add a user-agent header to the HTTP exporter via the headers config. This change preserves this behavior, and only sets the collector user-agent, if one is not present in config headers.
    Currently, it's not possible to do the above with a gRPC exporter. This change does not affect that, and only updates the existing gRPC client user-agent.

My understanding is that this is currently possible for the grpc exporter via the `headers` configuration and and **not** for the http exporter, but the comment in the description makes me think it's the other way around. Can you confirm this?

**Comment by vreynolds** [comment]
> My understanding is that this is currently possible for the grpc exporter via the `headers` configuration and and **not** for the http exporter, but the comment in the description makes me think it's the other way around. Can you confirm this?

Confirmed in local testing: `http` exporter allows setting user-agent via headers config, `grpc` does *not*

- For the `http` client, config headers are set as request headers, which is how a user-agent gets added canonically.
- For the `grpc` client, config headers are set as metadata, whereas a user-agent needs to be set as a dial option.

**Comment by codeboten** [comment]
> > My understanding is that this is currently possible for the grpc exporter via the `headers` configuration and and **not** for the http exporter, but the comment in the description makes me think it's the other way around. Can you confirm this?
> 
> Confirmed in local testing: `http` exporter allows setting user-agent via headers config, `grpc` does _not_
> 
>     * For the `http` client, config headers are set as request headers, which is how a user-agent gets added canonically.
> 
>     * For the `grpc` client, config headers are set as metadata, whereas a user-agent needs to be set as a dial option.

Thanks for clarifying 👍

File: exporter/otlpexporter/otlp.go
Comment (bogdandrutu) [reply exporter/otlpexporter/otlp.go:41]: How would that improve things?

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:123]: Please comment on how to use this, also a bit strange that is not an option in the "ToClient" but a func that changes the internal state.

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:134]: I would say that this is not a "clean" pattern, we should not change the user's config if possible.

File: exporter/otlphttpexporter/otlp.go
Comment (bogdandrutu) [inline exporter/otlphttpexporter/otlp.go:88]: Should we make this user configuration? Not sure if this should be enforced.

File: config/confighttp/confighttp.go
Comment (vreynolds) [reply config/confighttp/confighttp.go:123]: See https://github.com/open-telemetry/opentelemetry-collector/pull/3970#discussion_r702134406

Adding it to the ToClient interface was my initial thought as well, though I also see why changing that interface might be undesirable.

File: config/confighttp/confighttp.go
Comment (vreynolds) [reply config/confighttp/confighttp.go:134]: Hmm, I see what you're saying. How about adding the default user-agent as part of the default client settings? https://github.com/open-telemetry/opentelemetry-collector/blob/1d4b414590f9b41377df560a452638cbf6358732/exporter/otlphttpexporter/factory.go#L45

Comment (vreynolds) [reply config/confighttp/confighttp.go:134]: (assuming the build info is available at that point in the lifecycle)

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go:88]: Currently, the user-agent header for HTTP and gRPC exporters falls back to their respective defaults (e.g. `Go-http-client/1.1` `grpc-go/1.40.0`). Users can't opt out of that. (Though they can override it for HTTP requests - https://github.com/open-telemetry/opentelemetry-collector/pull/3970#issuecomment-916372313)

To me, this change updates that default fallback header to be more specific, which doesn't seem like it ought to be configurable, beyond the existing behavior of users providing a custom user-agent header.

WDYT?

**Review by codeboten — APPROVED** [review]
PR looks good, please resolve the conflict.

**Comment by vreynolds** [comment]
Merged upstream, and a couple small updates adding a doc comment and parenthesis around (OS/ARCH)

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [reply config/confighttp/confighttp.go:124]: I think we should start adding "Options" to that func and this can be one of the first option we create.

Also, this can be done by the caller already since they have access to the "Headers", how useful is this? How many callers do we have? 

I am trying to limit the API unless we have a clear reason to have it.

File: config/confighttp/confighttp.go
Comment (tigrannajaryan) [reply config/confighttp/confighttp.go:135]: +1 to what @bogdandrutu said. 

The reason for that is because we want to be able to show user-defined config exactly as it is (e.g. in zpages) and modifying it here will set a bad precedent that what we show is not exactly what the user defined, but may be different from it. I think it is undesirable.

Perhaps a better place to inject the default value for "User-Agent" header is in ToClient() func. Check that it is not specified by the user and add the default.

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [reply config/confighttp/confighttp.go:124]: Adding options to `ToClient` would also be a way to handle this.  I initially proposed a separate method to avoid changing the existing interface at all, but adding a variadic parameter is as close to no change as you can get since it should not affect existing callers.

File: config/confighttp/confighttp.go
Comment (Aneurysm9) [reply config/confighttp/confighttp.go:135]: > I would say that this is not a "clean" pattern, we should not change the user's config if possible.

I don't see this as changing the user's config any more than it would be if we did this as an option on `ToClient`.  If the user provides a user agent header, it is used, otherwise a default specified by the component using the HTTP client is used.  As for zpages, it might be odd that it wouldn't reflect this configuration until first use, but I think having it reflect the actual configuration is useful.

File: exporter/otlphttpexporter/otlp.go
Comment (Aneurysm9) [reply exporter/otlphttpexporter/otlp.go:88]: Indeed, I'd go further and say that this default user agent should always be used by the HTTP and gRPC clients provided by the collector unless an alternative is explicitly set by the user, not just in the exporters that opt into it.  I think that might be complicated by the lack of some of the required information in the clients, `BuildInfo` in particular.

**Comment by vreynolds** [comment]
Didn't get a chance to review the feedback last week, but will do that this week 🙏

File: config/confighttp/confighttp.go
Comment (vreynolds) [reply config/confighttp/confighttp.go]: https://github.com/open-telemetry/opentelemetry-collector/pull/3970#issuecomment-936850553

File: config/confighttp/confighttp.go
Comment (vreynolds) [reply config/confighttp/confighttp.go]: https://github.com/open-telemetry/opentelemetry-collector/pull/3970#issuecomment-936850553

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go:234]: Are you referring to the request modification? This is exactly what the HTTP client is doing currently: https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/confighttp.go#L124 (not to say that it makes it valid)

The compression roundtripper (same method, below) makes a copy of the request, then modifies the body and headers. Would that be acceptable here?

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go:88]: Noted, that sounds like exactly what I want here.

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go:221]: This type will be removed in favor of CustomRoundTripper

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
+1 to the user-agent idea, as I think this will be useful for other vendors as well. LGTM once the technical details & feedback are addressed

**Comment by bogdandrutu** [comment]
@vreynolds this is on you right? you still need to do the changes.

**Comment by vreynolds** [comment]
@bogdandrutu yes, I plan on updating this today -- are you OK with a custom roundtripper that clones the request and updates the header?

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go]: https://github.com/open-telemetry/opentelemetry-collector/pull/3970/commits/0b44ddff487109deb22b02730c9d9fdc4cee4141

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go]: Turns out I did need this as a receiver still -- but it is private now
https://github.com/open-telemetry/opentelemetry-collector/pull/3970/commits/0b44ddff487109deb22b02730c9d9fdc4cee4141

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go]: https://github.com/open-telemetry/opentelemetry-collector/pull/3970/commits/0b44ddff487109deb22b02730c9d9fdc4cee4141

File: exporter/exporterhelper/common.go
Comment (bogdandrutu) [inline exporter/exporterhelper/common.go:239]: let's move this to otlpexporter for the moment or an internal package. Prefer to limit the public API.

File: exporter/otlphttpexporter/otlp.go
Comment (bogdandrutu) [inline exporter/otlphttpexporter/otlp.go:222]: Do we really need a roundtripper? Since this is private, we should just simply add a header to the request when creating the request instead of cloning etc., as we need to do for a proper roundtripper.

**Review by bogdandrutu — APPROVED** [review]
few more comments about limiting the public API, and avoid implementing roundtripper if we don't intend that to be used by others.

File: exporter/exporterhelper/common.go
Comment (vreynolds) [reply exporter/exporterhelper/common.go]: ✅ Moved to each exporter https://github.com/open-telemetry/opentelemetry-collector/pull/3970/commits/f5bddaf5d217f650fa1bbd13990e43318c15a3d2

File: exporter/otlphttpexporter/otlp.go
Comment (vreynolds) [reply exporter/otlphttpexporter/otlp.go]: I did consider just adding the header when request is created, which is easier but more error-prone (need to remember to add the header again, if there's ever a new codepath that creates requests). That being said, I expect the exporter is fairly stable at this point, so this should be OK? https://github.com/open-telemetry/opentelemetry-collector/pull/3970/commits/f5bddaf5d217f650fa1bbd13990e43318c15a3d2

**Comment by vreynolds** [comment]
@bogdandrutu let me know if there's anything I need to do about the codecov report -- as far as I can tell, coverage is actually up, so I'm not sure what the failure is telling me

**Comment by jpkrohling** [comment]
> I'm not sure what the failure is telling me

It's not a problem with this PR. It's just telling that the project as a whole is under the target of 90%, which has been the case since we brought the builder here.

**Comment by codeboten** [comment]
Closing and re-opening to kick the CI

**Comment by codeboten** [comment]
@open-telemetry/collector-maintainers please merge once CI passes, looks like we have all the approvals needed

**Comment by jpkrohling** [comment]
The new release created a conflict for this PR. Sorry about that, but could you rebase the changelog, @vreynolds?

**Comment by vreynolds** [comment]
@jpkrohling conflict resolved ✅

**Review by jpkrohling — APPROVED** [review]

**Comment by robbkidd** [comment]
So excited! Can't wait!

**Comment by codeboten** [comment]
@tigrannajaryan @bogdandrutu please merge 😬

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
